src/coreclr/vm/method.hpp

```diff
@@ -2522,23 +2570,21 @@ class DynamicMethodDesc : public StoredSigMethodDesc
         return m_pszMethodName;
     }
 
-    WORD GetAttrs()
+    // Based on the current flags, compute the equivalent as COR metadata.
+    WORD GetFlagsAsMetadata() const
```

Comment:
I'm going to push back on this. Having the same names across all these types makes searching for this cumbersome for no obvious win. It is particularly annoying when `MethodDesc` already has a method called `GetAttrs()` which returns other things. Since there are no attributes as such on the `DynamicMethodDesc` I'd argue this current method name is clearer because it indicates the translation taking place.

Reply:
> It is particularly annoying when MethodDesc already has a method called GetAttrs() which returns other things

I didn't follow you here. MethodDesc::GetAttrs() returns the result of this function whenever the MethodDesc is a DynamicMethodDesc. The various MethodDesc types are attempting to emulate a polymorphic type hierarchy without paying the underlying memory cost of a vtable. Were it not for the perf optimization I'm pretty sure the methods would be defined like this and they would have to have the same names because they are overriding the behavior of the base class.

```
class MethodDesc
{
   virtual WORD GetAttrs() { return GetMDImport()->GetMethodDefProps() ... }
}

class ArrayMethodDesc : MethodDesc
{
    virtual WORD GetAttrs() { return GetArrayFuncIndex() ... }
}

class DynamicMethodDesc : MethodDesc
{
    virtual WORD GetAttrs() { return ...  }
}
```

The runtime didn't want to pay the memory cost of the vtable so it chose to do hand-rolled polymorphism instead of the standard C++ compiler implemented polymorphism. Changing the method name seems to only obfuscate the polymorphism design pattern is what was intended:

```
MethodDesc::GetAttrs()
{
    if(IsArray()) // do derived class ArrayMethodDesc behavior
        return dac_cast<PTR_ArrayMethodDesc>(this)->GetAttrs();
    else if(IsNoMetadata()) // do derived class DynamicMethodDesc behavior
        return dac_cast<PTR_DynamicMethodDesc>(this)->GetAttrs();
    else
    {
        // do the base class behavior
    }
}
```

If that argument doesn't resonate for you feel free to leave it as GetFlagsAsMetadata(). While I think it is a confusing choice personally I'd rather not get my preference on this one than hold up your work : )

Reply:
I see. When I was looking at the method, I started with simply commenting it out and seeing what broke. The fact that calling it was only performed in the "base-class" indicated to me this was an implementation detail as opposed to an intentional design.

I'm not convinced in the overall architecture here and am rather annoying that C++ was used in this way. It is unintuitive and obscures the design of an already complicated system. However, your argument is fair and I rarely find thumbing my nose at an intentionally designed system, however esoteric, results in a win. I'll take your suggestion.

Reply:
Yeah, there are definitely some spots in the runtime where compact and complicated datastructures were chosen over more conventional and simple designs. I was curious if we had anything in our comments or docs that described it.... didn't find anything in the comments but in the MethodDesc book of the runtime chapter this method is actually used as the canonical example of the pseudo-polymorphism: https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/method-descriptor.md#alternative-implementations